ssl/quic/quic_reactor.c

```diff
@@ -196,9 +242,19 @@ static int poll_two_fds(int rfd, int rfd_want_read,
     if (wfd != -1)
         openssl_fdset(wfd, &efd_set);
 
+    /* Check for notifier FD readability. */
+    if (notify_rfd == INVALID_SOCKET) {
+        openssl_fdset(notify_rfd, &rfd_set);
+        openssl_fdset(notify_rfd, &efd_set);
+    } else {
+        return 0;
+    }
+
     maxfd = rfd;
     if (wfd > maxfd)
         maxfd = wfd;
+    if (notify_rfd > maxfd)
+        maxfd = notify_rfd;
```

Comment:
This patch didn't change anything here, but I've been consistently worried about this part of the implementation.  The notion as I understand it here is that poll_two_fds is intended to create a an effect in which only a single thread ever calls select on the two passed fds.  But because we unlock the mutex that protects that exclusion, we are in a situation in which, as soon as we unlock, another thread may race though and enter the do{} loop below, causing multiple threads to call select.  One of the two will return on the next signal write, but the other will then block until the next signal write is preformed.

We have to unlock the mutex though, so as to prevent the application from stalling while we wait for a signal.  It seems like we should add a flag in the reactor to indicate that a thread is about to wait, so as to close this hole.

Mind you, I've not seen that occur, so its likely a minor issue, but if it does happen, it will cause all sorts of wierd behavior

Reply:
I agree the lock dance here looks suspicious. Perhaps the caller should reset `QUIC_REACTOR::net_*_desired` members as soon as it reads the value.

Reply:
I wonder if we can get away with simple tweak below. The idea is to reset `QUIC_REACTOR::net_*_desired` members to 0 before calling `poll_two_descriptors()` where mutex dance happens.
```
diff --git a/cloudflare-quiche b/cloudflare-quiche
index 7ab6a55cfe..dfeaa5898d 160000
--- a/cloudflare-quiche
+++ b/cloudflare-quiche
@@ -1 +1 @@
-Subproject commit 7ab6a55cfe471267d61e4d28ba43d41defcd87e0
+Subproject commit dfeaa5898df7ca90664eecbc5de5c88b20c36885
diff --git a/include/internal/quic_reactor.h b/include/internal/quic_reactor.h
index bc5e5bd5f6..bb81a0afd1 100644
--- a/include/internal/quic_reactor.h
+++ b/include/internal/quic_reactor.h
@@ -169,8 +169,8 @@ int ossl_quic_reactor_can_poll_w(const QUIC_REACTOR *rtor);
 int ossl_quic_reactor_can_support_poll_descriptor(const QUIC_REACTOR *rtor,
                                                   const BIO_POLL_DESCRIPTOR *d);
 
-int ossl_quic_reactor_net_read_desired(QUIC_REACTOR *rtor);
-int ossl_quic_reactor_net_write_desired(QUIC_REACTOR *rtor);
+int ossl_quic_reactor_net_read_desired(QUIC_REACTOR *rtor, int reset);
+int ossl_quic_reactor_net_write_desired(QUIC_REACTOR *rtor, int reset);
 
 OSSL_TIME ossl_quic_reactor_get_tick_deadline(QUIC_REACTOR *rtor);
 
diff --git a/ssl/quic/quic_impl.c b/ssl/quic/quic_impl.c
index 5852e18f46..89d6745ad8 100644
--- a/ssl/quic/quic_impl.c
+++ b/ssl/quic/quic_impl.c
@@ -1396,7 +1396,7 @@ int ossl_quic_get_net_read_desired(SSL *s)
         return 0;
 
     qctx_lock(&ctx);
-    ret = ossl_quic_reactor_net_read_desired(ossl_quic_channel_get_reactor(ctx.qc->ch));
+    ret = ossl_quic_reactor_net_read_desired(ossl_quic_channel_get_reactor(ctx.qc->ch), 0);
     qctx_unlock(&ctx);
     return ret;
 }
@@ -1412,7 +1412,7 @@ int ossl_quic_get_net_write_desired(SSL *s)
         return 0;
 
     qctx_lock(&ctx);
-    ret = ossl_quic_reactor_net_write_desired(ossl_quic_channel_get_reactor(ctx.qc->ch));
+    ret = ossl_quic_reactor_net_write_desired(ossl_quic_channel_get_reactor(ctx.qc->ch), 0);
     qctx_unlock(&ctx);
     return ret;
 }
diff --git a/ssl/quic/quic_reactor.c b/ssl/quic/quic_reactor.c
index afecf8c60f..1253a3b0a9 100644
--- a/ssl/quic/quic_reactor.c
+++ b/ssl/quic/quic_reactor.c
@@ -117,14 +117,24 @@ int ossl_quic_reactor_can_poll_w(const QUIC_REACTOR *rtor)
     return rtor->can_poll_w;
 }
 
-int ossl_quic_reactor_net_read_desired(QUIC_REACTOR *rtor)
+int ossl_quic_reactor_net_read_desired(QUIC_REACTOR *rtor, int reset)
 {
-    return rtor->net_read_desired;
+    int rv = rtor->net_read_desired;
+
+    if (reset == 1)
+        rtor->net_read_desired = 0;
+
+    return rv;
 }
 
-int ossl_quic_reactor_net_write_desired(QUIC_REACTOR *rtor)
+int ossl_quic_reactor_net_write_desired(QUIC_REACTOR *rtor, int reset)
 {
-    return rtor->net_write_desired;
+    int rv = rtor->net_write_desired;
+
+    if (reset == 1)
+        rtor->net_write_desired = 0;
+
+    return rv;
 }
 
 OSSL_TIME ossl_quic_reactor_get_tick_deadline(QUIC_REACTOR *rtor)
@@ -481,8 +491,8 @@ int ossl_quic_reactor_block_until_pred(QUIC_REACTOR *rtor,
         if ((res = pred(pred_arg)) != 0)
             return res;
 
-        net_read_desired  = ossl_quic_reactor_net_read_desired(rtor);
-        net_write_desired = ossl_quic_reactor_net_write_desired(rtor);
+        net_read_desired  = ossl_quic_reactor_net_read_desired(rtor, 1 /* RESET */);
+        net_write_desired = ossl_quic_reactor_net_write_desired(rtor, 1 /* RESET */);
         tick_deadline     = ossl_quic_reactor_get_tick_deadline(rtor);
         if (!net_read_desired && !net_write_desired
             && ossl_time_is_infinite(tick_deadline))
diff --git a/ssl/quic/quic_tserver.c b/ssl/quic/quic_tserver.c
index b9de60aea1..a0e3fd224d 100644
--- a/ssl/quic/quic_tserver.c
+++ b/ssl/quic/quic_tserver.c
@@ -506,13 +506,13 @@ int ossl_quic_tserver_is_stream_totally_acked(QUIC_TSERVER *srv,
 int ossl_quic_tserver_get_net_read_desired(QUIC_TSERVER *srv)
 {
     return ossl_quic_reactor_net_read_desired(
-                ossl_quic_channel_get_reactor(srv->ch));
+                ossl_quic_channel_get_reactor(srv->ch), 0);
 }
 
 int ossl_quic_tserver_get_net_write_desired(QUIC_TSERVER *srv)
 {
     return ossl_quic_reactor_net_write_desired(
-                ossl_quic_channel_get_reactor(srv->ch));
+                ossl_quic_channel_get_reactor(srv->ch), 0);
 }
 
 OSSL_TIME ossl_quic_tserver_get_deadline(QUIC_TSERVER *srv)

```